datafusion/src/physical_plan/expressions/binary.rs

```diff
@@ -878,26 +936,92 @@ impl PhysicalExpr for BinaryExpr {
     }
 }
 
+/// The binary_array_op_scalar macro includes types that extend beyond the primitive,
+/// such as Utf8 strings.
+#[macro_export]
+macro_rules! binary_array_op_dyn_scalar {
+    ($LEFT:expr, $RIGHT:expr, $OP:ident) => {{
+        let is_numeric = DataType::is_numeric($LEFT.data_type());
+        let is_numeric_dict = match $LEFT.data_type() {
+            DataType::Dictionary(_, val_type) => DataType::is_numeric(val_type),
+            _ => false
+        };
+        let numeric_like = is_numeric | is_numeric_dict;
+
+        let is_string = ($LEFT.data_type() == &DataType::Utf8) | ($LEFT.data_type() == &DataType::LargeUtf8);
```

Comment:
It seems to me that the `dyn` kernel to pick should be based on the type of `RIGHT` (the scalar` rather than `LEFT` (as there is already a pile of dispatch logic in the `eq_dyn` kernels)

I wonder if we could do something like

```rust
match $RIGHT {
  ScalarValue::Utf8(v) => {
    Ok(Arc::new(paste::expr! {[<$OP _dyn_scalar_utf8>]}(
            $LEFT,
            v,
        )?))
  }
  ..
  ScalarValue::Int8(v) => {
        Ok(Arc::new(paste::expr! {[<$OP _dyn_scalar>]}(
            $LEFT,
            v,
        )?))
  }
  ...
}
```

Though we will probably need some sort of wrapper to handle types not yet supported in arrow-rs 🤔

Reply:
@alamb this basically solved everything i think...we'll see how CI comes out